File: src/launcher/MinecraftInstaller.java
package launcher;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class MinecraftInstaller {

    private static final String VERSION_MANIFEST_URL =
            "https://piston-meta.mojang.com/mc/game/version_manifest_v2.json";

    public interface StatusListener {
        void onStatus(String text, long done, long total);
    }

    private final Path mcDir;
    private final StatusListener listener;

    public MinecraftInstaller(Path mcDir, StatusListener listener) {
        this.mcDir = mcDir;
        this.listener = listener;
    }

    private void status(String text) {
        if (listener != null) listener.onStatus(text, 0, 0);
    }

    private void progress(String text, long done, long total) {
        if (listener != null) listener.onStatus(text, done, total);
    }

    /** Fetches the version manifest as a list of maps: id, type, url, releaseTime. */
    public static List<Map<String, Object>> fetchVersionList() throws IOException, InterruptedException {
        String body = Http.getString(VERSION_MANIFEST_URL);
        Map<String, Object> root = Json.parseObject(body);
        List<Object> versions = Json.asArray(root.get("versions"));
        List<Map<String, Object>> out = new ArrayList<>();
        for (Object v : versions) {
            out.add(Json.asObject(v));
        }
        return out;
    }

    /** Returns the list of already-installed version IDs (based on versions/<id>/<id>.json presence). */
    public List<String> getInstalledVersions() {
        List<String> out = new ArrayList<>();
        Path versionsDir = mcDir.resolve("versions");
        if (!Files.isDirectory(versionsDir)) return out;
        try {
            Files.list(versionsDir).forEach(dir -> {
                if (Files.isDirectory(dir)) {
                    String id = dir.getFileName().toString();
                    if (Files.exists(dir.resolve(id + ".json"))) {
                        out.add(id);
                    }
                }
            });
        } catch (IOException ignored) {
        }
        return out;
    }

    /**
     * Installs a vanilla version: version json, client jar, libraries + natives, assets.
     * Returns the parsed version json map (also handed to the launcher).
     */
    public Map<String, Object> installVersion(String versionId) throws Exception {
        status("Resolving version " + versionId + "...");
        List<Map<String, Object>> versions = fetchVersionList();
        Map<String, Object> match = null;
        for (Map<String, Object> v : versions) {
            if (versionId.equals(v.get("id"))) {
                match = v;
                break;
            }
        }
        if (match == null) {
            throw new IOException("Version not found in manifest: " + versionId);
        }

        Path versionDir = mcDir.resolve("versions").resolve(versionId);
        Files.createDirectories(versionDir);

        String versionJsonUrl = Json.asString(match.get("url"), null);
        status("Downloading version metadata...");
        String versionJsonText = Http.getString(versionJsonUrl);
        Files.writeString(versionDir.resolve(versionId + ".json"), versionJsonText, StandardCharsets.UTF_8);
        Map<String, Object> versionJson = Json.parseObject(versionJsonText);

        installFromVersionJson(versionId, versionJson);
        return versionJson;
    }

    /** Shared install logic also used by Fabric (which reuses the vanilla parent's assets/libs). */
    @SuppressWarnings("unchecked")
    public void installFromVersionJson(String versionId, Map<String, Object> versionJson) throws Exception {
        Path versionDir = mcDir.resolve("versions").resolve(versionId);
        Files.createDirectories(versionDir);

        // --- client jar ---
        Map<String, Object> downloads = Json.asObject(versionJson.get("downloads"));
        Map<String, Object> clientDownload = Json.asObject(downloads.get("client"));
        String clientUrl = Json.asString(clientDownload.get("url"), null);
        Path clientJar = versionDir.resolve(versionId + ".jar");
        if (clientUrl != null && !Files.exists(clientJar)) {
            status("Downloading client jar...");
            Http.download(clientUrl, clientJar, (done, total) -> progress("Client jar", done, total));
        }

        // --- libraries + natives ---
        Path librariesDir = mcDir.resolve("libraries");
        Path nativesDir = versionDir.resolve(versionId + "-natives");
        Files.createDirectories(nativesDir);

        List<Object> libraries = Json.asArray(versionJson.get("libraries"));
        int libIndex = 0;
        for (Object lo : libraries) {
            libIndex++;
            Map<String, Object> lib = Json.asObject(lo);
            if (!RuleUtil.isAllowed(lib)) continue;

            Map<String, Object> libDownloads = Json.asObject(lib.get("downloads"));
            Map<String, Object> artifact = Json.asObject(libDownloads.get("artifact"));
            if (artifact.get("url") != null) {
                String url = Json.asString(artifact.get("url"), null);
                String path = Json.asString(artifact.get("path"), null);
                if (url != null && path != null) {
                    Path dest = librariesDir.resolve(path);
                    if (!Files.exists(dest)) {
                        status("Downloading library " + libIndex + "/" + libraries.size());
                        Http.download(url, dest, (done, total) -> progress("Libraries", done, total));
                    }
                }
            }

            // Legacy per-OS "classifiers" natives (pre 1.19 style)
            Object classifiersObj = libDownloads.get("classifiers");
            if (classifiersObj instanceof Map) {
                Map<String, Object> classifiers = (Map<String, Object>) classifiersObj;
                Map<String, Object> nativesMap = Json.asObject(lib.get("natives"));
                String classifierKey = Json.asString(nativesMap.get(RuleUtil.currentOsName()), null);
                if (classifierKey != null) {
                    classifierKey = classifierKey.replace("${arch}", RuleUtil.is64Bit() ? "64" : "32");
                    Object classifierEntryObj = classifiers.get(classifierKey);
                    if (classifierEntryObj != null) {
                        Map<String, Object> classifierEntry = Json.asObject(classifierEntryObj);
                        String url = Json.asString(classifierEntry.get("url"), null);
                        String path = Json.asString(classifierEntry.get("path"), null);
                        if (url != null && path != null) {
                            Path dest = librariesDir.resolve(path);
                            if (!Files.exists(dest)) {
                                status("Downloading natives...");
                                Http.download(url, dest, (done, total) -> progress("Natives", done, total));
                            }
                            extractNatives(dest, nativesDir);
                        }
                    }
                }
            }
        }

        // --- assets ---
        Object assetIndexObj = versionJson.get("assetIndex");
        if (assetIndexObj instanceof Map) {
            Map<String, Object> assetIndex = Json.asObject(assetIndexObj);
            String assetIndexId = Json.asString(assetIndex.get("id"), "legacy");
            String assetIndexUrl = Json.asString(assetIndex.get("url"), null);

            Path indexesDir = mcDir.resolve("assets").resolve("indexes");
            Files.createDirectories(indexesDir);
            Path indexFile = indexesDir.resolve(assetIndexId + ".json");

            String indexText;
            if (Files.exists(indexFile)) {
                indexText = Files.readString(indexFile, StandardCharsets.UTF_8);
            } else {
                status("Downloading asset index...");
                indexText = Http.getString(assetIndexUrl);
                Files.writeString(indexFile, indexText, StandardCharsets.UTF_8);
            }

            Map<String, Object> indexJson = Json.parseObject(indexText);
            Map<String, Object> objects = Json.asObject(indexJson.get("objects"));
            Path objectsDir = mcDir.resolve("assets").resolve("objects");

            int total = objects.size();
            int i = 0;
            for (Map.Entry<String, Object> entry : objects.entrySet()) {
                i++;
                Map<String, Object> obj = Json.asObject(entry.getValue());
                String hash = Json.asString(obj.get("hash"), null);
                if (hash == null) continue;
                String prefix = hash.substring(0, 2);
                Path dest = objectsDir.resolve(prefix).resolve(hash);
                if (!Files.exists(dest)) {
                    String url = "https://resources.download.minecraft.net/" + prefix + "/" + hash;
                    Http.download(url, dest, null);
                }
                if (i % 25 == 0 || i == total) {
                    progress("Assets " + i + "/" + total, i, total);
                }
            }
        }

        status("Install complete for " + versionId);
    }

    private void extractNatives(Path jarFile, Path targetDir) throws IOException {
        try (ZipInputStream zin = new ZipInputStream(Files.newInputStream(jarFile))) {
            ZipEntry entry;
            while ((entry = zin.getNextEntry()) != null) {
                String name = entry.getName();
                if (entry.isDirectory() || name.startsWith("META-INF/")) continue;
                // only extract native libs
                if (name.endsWith(".dll") || name.endsWith(".so") || name.endsWith(".dylib") || name.endsWith(".jnilib")) {
                    Path outPath = targetDir.resolve(new java.io.File(name).getName());
                    try (OutputStream out = Files.newOutputStream(outPath,
                            java.nio.file.StandardOpenOption.CREATE,
                            java.nio.file.StandardOpenOption.TRUNCATE_EXISTING)) {
                        zin.transferTo(out);
                    }
                }
            }
        }
    }
}

File: electron/lib/minecraftInstaller.ts
import fs from 'fs';
import path from 'path';
import { getString, downloadWithRetry } from './http';
import { matches as sha1Matches } from './sha1Util';
import { isAllowed, currentOsName, is64Bit } from './ruleUtil';
import AdmZip from 'adm-zip';

const VERSION_MANIFEST_URL = 'https://piston-meta.mojang.com/mc/game/version_manifest_v2.json';

export type StatusListener = (text: string, done: number, total: number) => void;

export interface VersionEntry {
  id: string;
  type: string;
  url: string;
  releaseTime: string;
}

export class MinecraftInstaller {
  constructor(private mcDir: string, private listener?: StatusListener) {}

  private status(text: string) {
    this.listener?.(text, 0, 0);
  }

  private progress(text: string, done: number, total: number) {
    this.listener?.(text, done, total);
  }

  static async fetchVersionList(): Promise<VersionEntry[]> {
    const body = await getString(VERSION_MANIFEST_URL);
    const root = JSON.parse(body);
    return root.versions as VersionEntry[];
  }

  /** Returns the list of already-installed version IDs (based on versions/<id>/<id>.json presence). */
  getInstalledVersions(): string[] {
    const out: string[] = [];
    const versionsDir = path.join(this.mcDir, 'versions');
    if (!fs.existsSync(versionsDir)) return out;
    for (const id of fs.readdirSync(versionsDir)) {
      const dir = path.join(versionsDir, id);
      if (fs.statSync(dir).isDirectory() && fs.existsSync(path.join(dir, `${id}.json`))) {
        out.push(id);
      }
    }
    return out;
  }

  /** Deletes an installed version's folder (and its per-version profile dir) entirely. */
  deleteVersion(versionId: string): void {
    fs.rmSync(path.join(this.mcDir, 'versions', versionId), { recursive: true, force: true });
    fs.rmSync(path.join(this.mcDir, 'profiles', versionId), { recursive: true, force: true });
  }

  async installVersion(versionId: string): Promise<any> {
    this.status(`Resolving version ${versionId}...`);
    const versions = await MinecraftInstaller.fetchVersionList();
    const match = versions.find((v) => v.id === versionId);
    if (!match) throw new Error(`Version not found in manifest: ${versionId}`);

    const versionDir = path.join(this.mcDir, 'versions', versionId);
    fs.mkdirSync(versionDir, { recursive: true });

    this.status('Downloading version metadata...');
    const versionJsonText = await getString(match.url);
    fs.writeFileSync(path.join(versionDir, `${versionId}.json`), versionJsonText, 'utf-8');
    const versionJson = JSON.parse(versionJsonText);

    await this.installFromVersionJson(versionId, versionJson);
    return versionJson;
  }

  /** Shared install logic also used by Fabric (which reuses the vanilla parent's assets/libs). */
  async installFromVersionJson(versionId: string, versionJson: any): Promise<void> {
    const versionDir = path.join(this.mcDir, 'versions', versionId);
    fs.mkdirSync(versionDir, { recursive: true });

    // --- client jar ---
    const clientDownload = versionJson.downloads?.client;
    const clientUrl = clientDownload?.url;
    const clientSha1 = clientDownload?.sha1;
    const clientJar = path.join(versionDir, `${versionId}.jar`);
    if (clientUrl && !(await sha1Matches(clientJar, clientSha1))) {
      this.status('Downloading client jar...');
      await downloadWithRetry(clientUrl, clientJar, 3, (done, total) => this.progress('Client jar', done, total));
    }

    // --- libraries + natives ---
    const librariesDir = path.join(this.mcDir, 'libraries');
    const nativesDir = path.join(versionDir, `${versionId}-natives`);
    fs.mkdirSync(nativesDir, { recursive: true });

    const libraries: any[] = versionJson.libraries ?? [];
    let libIndex = 0;
    for (const lib of libraries) {
      libIndex++;
      if (!isAllowed(lib)) continue;

      const artifact = lib.downloads?.artifact;
      if (artifact?.url) {
        const url = artifact.url;
        const relPath = artifact.path;
        const sha1 = artifact.sha1;
        if (url && relPath) {
          const dest = path.join(librariesDir, relPath);
          if (!(await sha1Matches(dest, sha1))) {
            this.status(`Downloading library ${libIndex}/${libraries.length}`);
            await downloadWithRetry(url, dest, 3, (done, total) => this.progress('Libraries', done, total));
          }
        }
      }

      // Legacy per-OS "classifiers" natives (pre-1.19 style)
      const classifiers = lib.downloads?.classifiers;
      if (classifiers) {
        const nativesMap = lib.natives;
        let classifierKey: string | undefined = nativesMap?.[currentOsName()];
        if (classifierKey) {
          classifierKey = classifierKey.replace('${arch}', is64Bit() ? '64' : '32');
          const classifierEntry = classifiers[classifierKey];
          if (classifierEntry) {
            const url = classifierEntry.url;
            const relPath = classifierEntry.path;
            const sha1 = classifierEntry.sha1;
            if (url && relPath) {
              const dest = path.join(librariesDir, relPath);
              if (!(await sha1Matches(dest, sha1))) {
                this.status('Downloading natives...');
                await downloadWithRetry(url, dest, 3, (done, total) => this.progress('Natives', done, total));
              }
              this.extractNatives(dest, nativesDir);
            }
          }
        }
      }
    }

    // --- assets ---
    const assetIndex = versionJson.assetIndex;
    if (assetIndex) {
      const assetIndexId = assetIndex.id ?? 'legacy';
      const assetIndexUrl = assetIndex.url;

      const indexesDir = path.join(this.mcDir, 'assets', 'indexes');
      fs.mkdirSync(indexesDir, { recursive: true });
      const indexFile = path.join(indexesDir, `${assetIndexId}.json`);

      let indexText: string;
      if (fs.existsSync(indexFile)) {
        indexText = fs.readFileSync(indexFile, 'utf-8');
      } else {
        this.status('Downloading asset index...');
        indexText = await getString(assetIndexUrl);
        fs.writeFileSync(indexFile, indexText, 'utf-8');
      }

      const indexJson = JSON.parse(indexText);
      const objects: Record<string, any> = indexJson.objects ?? {};
      const objectsDir = path.join(this.mcDir, 'assets', 'objects');

      const virtual = !!indexJson.virtual;
      const mapToResources = !!indexJson.map_to_resources;
      const virtualDir = path.join(this.mcDir, 'assets', 'virtual', assetIndexId);
      const resourcesDir = path.join(this.mcDir, 'resources'); // legacy game dir layout

      const entries = Object.entries(objects);
      const total = entries.length;
      let i = 0;
      for (const [assetPath, obj] of entries as [string, any][]) {
        i++;
        const hash = obj.hash;
        if (!hash) continue;
        const prefix = hash.slice(0, 2);
        const dest = path.join(objectsDir, prefix, hash);
        if (!(await sha1Matches(dest, hash))) {
          const url = `https://resources.download.minecraft.net/${prefix}/${hash}`;
          await downloadWithRetry(url, dest, 3, null);
        }

        // Very old versions (pre-1.7) expect assets laid out by their
        // real path rather than by hash - materialize that layout too.
        if (virtual) this.copyIfMissing(dest, path.join(virtualDir, assetPath));
        if (mapToResources) this.copyIfMissing(dest, path.join(resourcesDir, assetPath));

        if (i % 25 === 0 || i === total) this.progress(`Assets ${i}/${total}`, i, total);
      }
    }

    this.status(`Install complete for ${versionId}`);
  }

  private copyIfMissing(source: string, dest: string) {
    try {
      if (fs.existsSync(source) && !fs.existsSync(dest)) {
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.copyFileSync(source, dest);
      }
    } catch {
      /* ignored */
    }
  }

  private extractNatives(jarFile: string, targetDir: string): void {
    const zip = new AdmZip(jarFile);
    for (const entry of zip.getEntries()) {
      const name = entry.entryName;
      if (entry.isDirectory || name.startsWith('META-INF/')) continue;
      if (/\.(dll|so|dylib|jnilib)$/.test(name)) {
        const base = path.basename(name);
        fs.mkdirSync(targetDir, { recursive: true });
        fs.writeFileSync(path.join(targetDir, base), entry.getData());
      }
    }
  }
}
